### PR TITLE
Log-conformation formulation for viscoelastic flows

### PR DESCRIPTION
**Pull request by richmartin33**
Adding the log-conformation formulation to the viscoelastic capabilities of Goma.

### CODE REVIEW

File: src/mm_fill.c [hunk 1]
```diff
@@ -3169,6 +3181,1657 @@ matrix_fill(
...
+	  	{
+	          fprintf(stderr,"CA:%d ID:%d fselem:%d sselem:%d Proc:%d\n",j,CA_id[j],CA_fselem[j],CA_sselem[j],CA_proc[j]);
+		}
+	    fprintf(stderr,"Count=%d  Done=%d\n",count,Num_CAs_done);
+	  }
+      }
+    }
+
+  return 0;
+} /*   END OF matrix_fill_stress                                                     */
```
Comment (wortiz) [inline src/mm_fill.c]: I don't think there is a better way to do this without rewriting matrix fill to be more modular. But it is very ugly to have a ton of duplicated code for only filling in stress.

File: src/mm_fill_stress.c [hunk 2]
```diff
@@ -5222,3 +5659,167 @@ stress_eqn_pointer(int v_s[MAX_MODES][DIM][DIM])
...
+      for (p = 0; p < VIM; p++) {
+	for (q = 0; q < VIM; q++) {
+	  d_exp_s_ds[p][q][i][j] = (exp_s_p[p][q] - exp_s[p][q]) / ds;
+	}
+      }
+      
+      for (m = 0; m < VIM; m++) {
+        for (n=0; n < VIM; n++) {
+          s_p[m][n] = s[m][n];
+        }
```
Comment (wortiz) [inline src/mm_fill_stress.c]: You can save a small amount of computation here by just undoing what was perturbed

e.g

    s_p[i][j] += s[i][j]
    if( i != j) {
      s_p[j][i] = s[j][i]
    }

File: src/mm_fill_terms.c [hunk 3]
```diff
@@ -4024,8 +4031,19 @@ assemble_momentum(dbl time,       /* current time */
 					  {
 					    for ( q=0; q<VIM; q++)
 					      {
-						diffusion -=
-						  grad_phi_i_e_a[p][q] * d_Pi->S[p][q][mode][b][c][j];
+                                                if(vn->evssModel == LOG_CONF)
+				                  {
+				                    if( b <= c )
```
Comment (wortiz) [inline src/mm_fill_terms.c]: Why this is needed should be figured out before this is merged

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by acochrane** [comment]
.runguts

**Comment by wortiz** [comment]
.runguts

**Comment by wortiz** [comment]
As stress variable is being used here as log_conf, we should either switch to using another variable (which might be troublesome because of how modes are implemented), or perform a review and update boundary conditions/equations or give unimplemented warnings when running with log conf.

**Pull request merged**